src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryAccess.java

```diff
@@ -2031,6 +2033,15 @@ public static MemoryAddress getAddressAtIndex(MemoryAddress addr, long index) {
      * @param value the memory address to be written.
      */
     public static void setAddressAtIndex(MemoryAddress addr, long index, MemoryAddress value) {
-        setAddressAtOffset(addr, index * 8, value);
+        setAddressAtOffset(addr, scale(index, 8), value);
+    }
+
+    @ForceInline
+    private static long scale(long index, int size) {
+        if (index < Integer.MAX_VALUE) {
+            return (int)index * size;
+        } else {
+            return index * size;
+        }
     }
```

Comment:
This looks problematic. `index * size` can overflow when casting to int, when it would not have overflown if the operands were `long`. For small segments we do something similar, but it's less of a problem there since we know the segment size fits in an int, so if there's overflow the access would have been out of bounds any ways. I think here you will have to pass in the `segment()` from the accessed address, and then use the isSmall() predicate as the criteria for doing the cast or not.

Reply:
Maybe MemoryAddressProxy::multiplyOffsets can be used?